### DISCUSSION TITLE
【主库】The ObjectPool has been disposed

### DISCUSSION THREAD

**Discussion by 2881099** [dotnetcore/FreeSql#1079] [Q&A]
### 【主库】The ObjectPool has been disposed

---

原因一：手工调用了 fsql.Dispose，之后仍然使用它

---

原因二：使用了 IdleBus 管理 IFreeSql，错误的方式如下：

- a) 不要构建 IFreeSql 之后再注册

```csharp
var fsql = new FreeSqlBulder()...Build();
ib.Register("key01", () => fsql); //错了，错了，错了

ib.Register("key01", () => new FreeSqlBulder()...Build()); //正确
```

- b) 尽量每次都使用 ib.Get 获得 IFreeSql 对象(避免存对象引用)，IdleBus 内部超时释机制一旦触发，再使用引用对象，就会报这个报错

```c#
var fsql = ib.Get("key01");
fsql.Select<T>().ToList(); //正常
Thread.CurrentThread.Join(TimeSpan.FromMinutes(10));
fsql.Select<T>().ToList(); //错误，因为 fsql 已经被 IdleBus 释放
```

---

原因三：检查项目的系统事件，是否在异常之前触发

```csharp
AppDomain.CurrentDomain.ProcessExit += (s1, e1) =>
{
  //记录日志
};
Console.CancelKeyPress += (s1, e1) =>
{
  //记录日志  
};
```

如果确定问题，可以在 FreeSqlBuilder 构建对象的时候 UseExitAutoDisposePool(false) 关闭这个机制

**Comment by chunjieyaya** [comment]
netcore8.0 ,在多线程中去执行语句，也报这个错

**Reply by 2881099** [reply]
如果使用了 Idlebus，把 IdleBus 换成 FreeSqlCloud

**Reply by chunjieyaya** [reply]
有相关的网址或者说明吗？谢谢大哥~

**Reply by 2881099** [reply]
文档里面搜索，多库

**Comment by pengxinglove** [comment]
![企业微信截图_17210939557683](https://github.com/user-attachments/assets/4a36b265-309c-4c16-ac64-d4bbc41f4358)
![image](https://github.com/user-attachments/assets/82298ffe-cdb5-4170-9ecf-bfc08fb927b4)
因为using(var customerdb = xxx)  导致freesql对象被释放了，如果有类似的问题，可以查看using相关语句
